make settings lazy-loaded when they are needed

emirotin opened the pull request:
I've come here the long way: `resin-request` -> `resin-settings-storage` -> `resin-settings-client` :)
`resin-settings-storage` tries to be universal, but there's a problem. It loads the `resin-settings-client` unconditionally (https://github.com/resin-io/resin-settings-storage/blob/master/lib/local-storage.coffee#L17), and when the module tries to do some initial fs manipulations it fails in the browser.
I should add that loading the module conditionally (moving the require inside of the `unless` clause) doesn't help. In my understanding the module is still included into the browserify bundle and its initialization code is executed. At least that's what I see when running the tests in PhantomJS, and I've checked that the `localStorage` is defined there.

So the change I've made here is to only load the settings once they're actually needed. Memoization makes the overhead minimal and should be barely noticeable in the CLI app.

Once this is merged and published I will send a PR to the `resin-settings-storage` that will use this change and add the browser testing.

I've also updated the docs on testing to adhere to the `npm` standard convention and hide the implementation details.

emirotin commented:
The test failed when trying to download node 0.12. How do we re-run it?

jviotti commented:
I see. Your workaround looks great to me, so I give the green flag to merge (I'll wait for you to merge any PR so you can have a better control of the order of the merged PRs, etc. Happy to publish for you if you like though, just ping me, and I'll publish any module as requested).

Regarding retrying to builds, you can click the "Details" link at the right of the failing builds, and the "Retry" button at the right side of the Travis CI page. Let me know if you don't see any of these and I'll see if there are any permissions issues.

As a side note, this modules were not updated since a long time, so the CI builds still run on ancient NodeJS versions, like v0.10. If you want, feel free to update the NodeJS versions in CI builds to something like v4 and v6, etc.

emirotin commented:
Cool, I'll update the node versions (will keep 0.12 though), and squash the commits.
I've already verified that this fix helps (together with some tweaks to the tests in the storage module).

I don't see the Retry button in Travis (I've checked it before as well), so probably only the collaborators can do that.

jviotti commented:
Sounds great! I'll get the collaborators issue sorted!

emirotin commented:
So, I've squashed the old commits. I've decided to keep the changes to the built library, I think it's OK here.
Updated Travis config to use 0.12, 4, and 6, all tests pass.
Updated AppVeyor accordingly (but it has disappeared from the checks in this repo).
And also the README was somehow deleted in your move PR, I've restored it here.

Please if you can restore the AppVeyor check and it passes merge this PR and publish a new version (strictly speaking it could be major as we drop support for older node versions, but I think it should actually be minor).

Page- commented on lib/settings.coffee:
```diff
@@ -70,22 +70,21 @@ config = require('./config')
-	readConfigFile(config.paths.project)
-	environment.parse(process.env)
-]
+getSettings = _.memoize ->
```
`_.once` is better in this situation

emirotin replied:
Indeed. Forgot about it.

Page- commented on lib/settings.coffee:
```diff
@@ -70,22 +70,21 @@ config = require('./config')
-	environment.parse(process.env)
-]
+getSettings = _.memoize ->
+	utils.mergeObjects.apply null, [
```
Do you know if this can be:

``` coffee
utils.mergeObjects(
    defaults
    readConfigFile(config.paths.user)
    readConfigFile(config.paths.project)
    environment.parse(process.env)
)
```

?

emirotin replied:
I've left it as it was. just wrapped in a function, but indeed was a bit surprised about the usage of `apply`.

Page- replied:
Yeah, it was more of a can we remove this somewhat confusing code, @jviotti do you know?

jviotti replied:
@emirotin @Page- I honestly don't remember. I do see that `lib/defaults.coffee` accesses `this` in some of the functions, so there might be a reason behind that. I tried removing and all tests still pass, so I guess it can be safely removed.
